### PR TITLE
[Feature] Add CUDA Weighted Randomwalk Sampling

### PR DESCRIPTION
**Pull request by zhenliangqiu**
## Description
Add CUDA Weighted Randomwalk sampling.

For a heterogeneous graph, the calculation process is as follows:

>- Traverse each CSR subgraph:   
>    - Use the CSR row offset and probability array to calculate the sum of the probabilities of all seed node neighbors W;  
>    -  Take a random number R;  
>    -  Calculate the random weight sum with W[i]*R;  
>    -  Traverse the neighbors of current seed in [0, degree) with n:  
>        - Accumulate the weights until the sum is greater than W[i]*R, so the n is the next sampling index.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes
Add two functions `RandomWalkBiased()` and `_RandomWalkBiasedKernel()` in file: `dgl/src/graph/sampling/randomwalks/randomwalk_gpu.cu` to support biased choice for random walk on GPU.

### CODE REVIEW

**Comment by mufeili** [comment]
Could you take a look? @yaox12

**Review by yaox12 — CHANGES_REQUESTED** [review]
Generally LGTM. Two comments:
1.  Can you enable GPU support for `RandomWalkWithRestart` and `RandomWalkWithStepwiseRestart` with probabilities? I think your kernel already covers them and only some host-side functions need to update.
2. Can you enable unit tests for biased random walk on GPU? It should be in `tests/compute/test_sampling.py`, and `test_non_uniform_random_walk()` more specifically.

Let @nv-dlasalle @lixiaobai09 have a look.

File: src/graph/sampling/randomwalks/randomwalk_gpu.cu [hunk 1]
```diff
@@ -167,23 +239,161 @@ std::pair<IdArray, IdArray> RandomWalkUniform(
...
+  // graphs
+  std::vector<GraphKernelData<IdType>> h_graphs(num_etypes);
+  for (int64_t etype = 0; etype < num_etypes; ++etype) {
+    const CSRMatrix &csr = hg->GetCSRMatrix(etype);
+    h_graphs[etype].in_ptr  = static_cast<const IdType*>(csr.indptr->data);
+    h_graphs[etype].in_cols = static_cast<const IdType*>(csr.indices->data);
+    h_graphs[etype].data = (CSRHasData(csr) ? static_cast<const IdType*>(csr.data->data) : nullptr);
+
+    int64_t num_segments = csr.indptr->shape[0] - 1;
+    probs[etype] = prob[etype].Ptr<FloatType>();
```
Comment (yaox12) [inline src/graph/sampling/randomwalks/randomwalk_gpu.cu:284]: `prob[etype]` can be empty.

**Review by zhenliangqiu — COMMENTED** [review]
@yaox12 Thank you for review and changes !  ^-^

**Review by yaox12 — APPROVED** [review]

**Comment by mfbalin** [comment]
What datasets were used to debug and test this PR? I am working on a new sampling algorithm PR #4668 as well, but I lack datasets that I can use to test my weighted sampling implementation.

**Pull request merged**